[openssl] remove run-time dependency of perl due to c_rehash

adrian-gierakowski opened the pull request:
###### Description of changes

Replaces perl based c_rehash script with shell script wrapping `openssl rehash`
with the same functionality.

Fixes: #19965
Supersedes: #156776, #83446
Possibly related to: #157093, #82924

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

adrian-gierakowski committed:
[openssl] remove run-time dependency of perl due to c_rehash

Replaces perl based c_rehash script with shell script wrapping `openssl rehash`
with the same functionality.

Fixes: #19965
Supersedes: #156776, #83446
Possibly related to: #157093, #82924

adrian-gierakowski commented:
as discussed [here](https://github.com/NixOS/nixpkgs/pull/156776#issuecomment-1453198061) the `withPerl` flag could be removed altogether

happy to do this if there are no objections

Izorkin commented:
Need to add these changes for `quictls`?

adrian-gierakowski commented:
> Need to add these changes for `quictls`?

Was just going to ask 😄

Once the final decision is made regarding  keeping\removing the withPerl flag I’ll apply the changes there as well

adrian-gierakowski commented:
@veprbl I saw you did code review on #156776 

Do you have the power to merge this? What's your opinion regarding removing the `withPerl` flag? Thanks!

Artturin commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -184,6 +191,13 @@ let
       # "#!/usr/bin/env perl"
     ''
       substituteInPlace $out/bin/c_rehash --replace ${buildPackages.perl}/bin/perl "/usr/bin/env perl"
+    '' + lib.optionalString (!withPerl)
+    # Replace perl based c_rehash script with c_rehash_shim
+    ''
+      substitute \
+        ${c_rehash_shim} \
+        $out/bin/c_rehash \
+        --replace /usr/bin/openssl "${placeholder "bin"}/bin/openssl"
```
```suggestion
    '' + lib.optionalString (!withPerl) ''
      makeWrapper $bin/bin/openssl $bin/bin/c_rehash \
        --add-flags "rehash"
```

Artturin commented:
let's drop withPerl

adrian-gierakowski committed:
[openssl] remove withPerl flag and simplify c_rehash shim

Co-authored-by: Artturi <Artturin@artturin.com>

adrian-gierakowski committed:
[quictls] remove run-time dependency of perl due to c_rehash

same as 2e1e4318e38526f0b18246f277dbc5be0b7641fc and e05a859566966597c47e8bfe4d7ed14712d087d6

adrian-gierakowski commented:
@Artturin removed `withPerl` and applied your improvement to the wrapper scripts
also applied the same changes to `quictls`

Artturin commented:
Needs to target staging and the commit msg's need to be changed to our style, instructions for both in `CONTRIBUTING.md`

thomasjm commented:
Hey, there's about a month until the 23.05 release and it would be awesome to get this merged before then. Are there any blockers remaining?

Artturin commented:
@ofborg build openssl

Artturin commented:
Needs to target staging check CONTRIBUTING.md for how to rebase

Artturin commented:
oh already said that

Artturin commented:
Can't push to this PR `ERROR: Permission to rhinofi/nixpkgs.git denied to Artturin.` even though `Maintainers are allowed to edit this pull request.`

Artturin commented:
https://github.com/NixOS/nixpkgs/pull/226108 kept your authorship

adrian-gierakowski commented:
@Artturin thanks! I didn’t really have time to address your comment so it’s great to took care of it :)
